BUG: signal.tf2sos: fix a new `ComplexWarning`

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

reported in https://github.com/scipy/scipy/issues/23221#issuecomment-2997264496

fixes gh-23221, fixes gh-23277

#### What does this implement/fix?
<!--Please explain your changes.-->

Cherry-pick a part of gh-22911 to fix an issue reported for scipy==1.16.0 in gh-23221

https://github.com/scipy/scipy/issues/23221#issuecomment-2997264496

#### Additional information
<!--Any additional information you think is important.-->

The PR is against maintenance/1.16.x since the problem is already fixed in main.

ev-br committed:
BUG: signal: more robust real check in _pu.poly

Is apparently needed on torch.

ev-br committed:
MAINT: signal: use a less surprising way to check if roots are complex conjugates

tylerjereddy reviewed (commented):
To confirm--is there no "regression" test? We can't check for absence of warning if that is the issue? Or the plan is to fix more correctly later (I thought I saw something about that in the ticket).

ev-br committed:
MAINT: signal: add a tf2sos test for no warnings

ev-br commented:
> To confirm--is there no "regression" test? We can't check for absence of warning if that is the issue?

OK, I added a test which will fail if the OP case from gh-23221 emits eny warnings.

>  Or the plan is to fix more correctly later (I thought I saw something about that in the ticket).

The issue is specific to scipy==1.6.0 and is already fixed in main.

For context: a series of array API PRs got split across the branching point; it turns out that what landed in 1.16.0 emits a warning in some cases; the fix is in main only; this PR backports the fix to maintenance/1.16.x.

ev-br committed:
BUG: signal: fix freqz for real/complex `w` arrays

Backport the relevant part of https://github.com/scipy/scipy/pull/22911
to fix https://github.com/scipy/scipy/issues/23277

ev-br commented:
Have added one more partial cherry-pick to fix https://github.com/scipy/scipy/issues/23277

tylerjereddy committed:
MAINT: PR 23222 revisions

* Fixed a SciPy version typo in the above PR.

[skip ci] [ci skip]

tylerjereddy commented:
I confirmed locally that both of the newly-added tests fail when the source changes in the PR are reverted.

There was a small typo in one of the tests talking about SciPy `1.6.0` instead of `1.16.0`. I pushed in a quick fix for that with CI skipped, since CI was already passing, and I'll now proceed with the backport merge as we're interested in getting moving with SciPy `1.16.1`.

Thanks Evgeni
